context.go

```diff
@@ -6,53 +6,142 @@ import (
 	"flag"
 	"fmt"
 	"strings"
+	"time"
 )
 
-// Context is a type that is passed through to
-// each Handler action in a cli application. Context
+// Context is an interface that is passed through to
+// each Handler action in a cli application. It
 // can be used to retrieve context-specific args and
 // parsed command-line options.
-type Context struct {
-	context.Context
-	App           *App
-	Command       *Command
+type Context interface {
+	// Functions to be able to manage Context type
+	App() *App
+	SetCommand(command *Command)
+	Command() *Command
+	Context() context.Context
+	ParentContext() Context
+	SetShellComplete(shellComplete bool)
+	ShellComplete() bool
+	SetFlagSet(set *flag.FlagSet)
+	FlagSet() *flag.FlagSet
+	NumFlags() int
+	Set(name, value string) error
+	IsSet(name string) bool
+	LocalFlagNames() []string
+	FlagNames() []string
+	Lineage() []Context
+	Value(name string) interface{}
+	Args() Args
+	NArg() int
+	// Functions for each flag type supported by the cli
```

Comment:
We probably want to move this into a separate Interface. The context interface is already pretty huge so we could think about 1-3 logical groups of interfaces:

- flag values
- app settings
- context relations (parent, etc.) - I’m not sure if we need this at all.

Reply:
Sounds good. Let me fix this up :)

Reply:
I have pushed new code with the `Context` interface split up.